napari/utils/progress.py

```diff
@@ -0,0 +1,245 @@
+}
+
+
+class progress(tqdm):
+    """This class inherits from tqdm and provides an interface for
+    progress bars in the napari viewer. Progress bars can be created
+    directly by wrapping an iterable or by providing a total number
+    of expected updates.
+
+    See tqdm.tqdm API for valid args and kwargs:
+    https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
+    ...         sleep(delay)
+
+    it can also be used as a context manager:
+
+    >>> def long_running(steps=10, repeats=4, delay=0.1):
+    ...     with progress(range(steps)) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    or equivalently, using the `progrange` shorthand
+    ...     with progrange(steps) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    For manual updates:
+
+    >>> def manual_updates(total):
+    ...     pbr = progress(total=total)
+    ...     sleep(10)
+    ...     pbr.set_description("Step 1 Complete")
+    ...     pbr.update(1)
+    ...     # must call pbr.close() when using outside for loop
+    ...     # or context manager
+    ...     pbr.close()
+
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+
+        # check if there's a napari viewer instance
+        viewer = get_viewer_instance()
+        self.has_viewer = viewer is not None
+        if self.has_viewer:
+            kwargs['gui'] = True
+
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        self.nested_token = None
+
+        super().__init__(iterable, desc, total, *args, **kwargs)
+        if not self.has_viewer:
+            return
+
+        self._pbar = get_pbar(viewer, **pbar_kwargs)
+        if self.total is not None:
+            self._pbar.setRange(self.n, self.total)
+            self._pbar._set_value(self.n)
+        else:
+            self._pbar.setRange(0, 0)
+
+        if desc:
+            self.set_description(desc)
+        else:
+            desc = get_calling_function_name(max_depth=5)
+            if desc:
+                self.set_description(desc)
+            else:
+                # TODO: pick a better default
+                self.set_description("Progress Bar")
+
+        self.show()
+        QApplication.processEvents()
+
+    def __enter__(self):
+        self.nested_token = IS_NESTED.set(True)  # noqa
+        return super().__enter__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if IS_NESTED.get():
+            IS_NESTED.reset(self.nested_token)
```

Comment:
unless someone is abusing the API, `__exit__` will likely only be called after `__enter__` has been called.  so you can a) set `self._nested_token` (make it private) only on the `__enter__` method, and just reset it without checking in `__exit__`.

```suggestion
    def __enter__(self):
        self._nested_token = CURRENT_PBAR.set(id(self))
        return super().__enter__()

    def __exit__(self, exc_type, exc_value, traceback):
        CURRENT_PBAR.reset(self._nested_token)
```